backends/ebpf/psa/externs/ebpfPsaRegister.cpp

```diff
@@ -161,27 +187,33 @@ void EBPFRegisterPSA::emitRegisterRead(CodeBuilder* builder, const P4::ExternMet
     cstring indexParamName = translator->getParamName(indexArg);
     BUG_CHECK(!indexParamName.isNullOrEmpty(), "Index param cannot be empty");
 
-    cstring valueName = program->refMap->newName("value");
+    this->readValueName = program->refMap->newName("value");
 
     builder->emitIndent();
-    this->valueType->declare(builder, valueName, true);
+    this->valueType->declare(builder, readValueName, true);
     builder->endOfStatement(true);
 
     cstring msgStr = Util::printf_format("Register: reading %s",
                                  instanceName.c_str());
     builder->target->emitTraceMessage(builder, msgStr.c_str());
 
     builder->emitIndent();
-    builder->target->emitTableLookup(builder, dataMapName, indexParamName, valueName);
+    builder->target->emitTableLookup(builder, dataMapName, indexParamName, readValueName);
     builder->endOfStatement(true);
 
     builder->emitIndent();
-    builder->appendFormat("if (%s != NULL) ", valueName.c_str());
+    builder->appendFormat("if (%s != NULL) ", readValueName.c_str());
     builder->blockStart();
+    if (this->isAtomic) {
```

Comment:
The spinlock is protecting a region between `register.read()` and `register.write()`. Other cases are not possible to protect because of spinlocks constraints.

Reply:
What if you have multiple reads followed by a write?

Reply:
When you have a register is it shared across all cores?